Js edit entries and index page

johnschoeman opened the pull request:
This allows users to edit already created entries.

johnschoeman committed:
Get first pass of edit entry feature working

This gets the edit entry feature spec to pass, but has not been
refactored yet. The way that the form object was written 'works' but was
not well suited to be expanded to include new functionality.  I want to
save this intermediate step so that i can reference (potential good blog
post) why it is sub-ideal and compare to the better pattern shown after
the refactor.

johnschoeman committed:
Get second pass of edit entry feature working

This refactors the first pass, but doesn't add any new functionality. It
is better but still not great, there are some issues with the scope of
what the form object is doing. I want to keep this intermediate stage
around for reference on bad and better ways to do form objects.

johnschoeman commented on app/models/entry.rb:
```diff
@@ -3,6 +3,14 @@ class Entry < ApplicationRecord
    validates :date, presence: true
 
-  has_many :goals, dependent: :destroy
+  has_many :goals, -> { order(id: :asc) }, dependent: :destroy
```
I added this scope so that the ordering of the goals remained consistent, choosing to order by id was arbitrary.  A future feature could include allowing users to chose the order of the goals.

johnschoeman commented on app/models/entry_form.rb:
```diff
@@ -1,66 +1,90 @@
 
-  def persisted?
-    false
+  def self.find_by_entry_id(entry_id)
```
I couldn't find a more elegant pattern to get the EntryForm to 'repopulate itself' when doing the 'find_by_id' step in the edit/update flow.  Open to suggestions here.

christoomey replied:
My inclination would be to have the form initialize with an entry, either an `Entry.new` in the case of the `:new` action, or a found Entry in the case of the edit action.

johnschoeman commented on spec/features/user_edits_an_entry_spec.rb:
```diff
@@ -0,0 +1,57 @@
+
+    user = sign_in
+    entry_to_edit =
+      Entry.create(date: todays_date, goals: initial_goals, user: user)
```
I was having issues getting factory bot to persist entries in such a way that when the test goes to go edit them they recognize the already created entries.  So i went with just not using factory bot to get the tests to work correctly and will investigate how factory bot is meant to be used in this situation.

johnschoeman commented on spec/support/features/clearance_helpers.rb:
```diff
@@ -10,6 +10,7 @@ def sign_in
       password = "password"
       user = FactoryBot.create(:user, password: password)
       sign_in_with user.email, password
+      user
```
this was related to not being able to use factory bot fully in the edit an entry feature.  A bit of a hack to be able to make entries that belong to the signed in user.

johnschoeman commented:
I refactored the FormObject again, each iteration seems to be getting better! A next step would be to introduce a PageObject like we discussed the other day to improve the readability of the tests.

christoomey commented on app/models/entry_form.rb:
```diff
@@ -1,66 +1,90 @@
+        entry.save!
+        goals.each do |goal|
+          goal.save!
+        end
```
`goals.each(&:save!)` ?

christoomey commented on app/views/entries/edit.html.erb:
```diff
@@ -0,0 +1,22 @@
+    />
+  <% end %>
+  <%= f.submit "Submit Edit Entry" %>
+<% end %>
```
In general we strive to re-use the same form for new / edit. Can you extract the form from the `new` view and reuse that here?

christoomey commented on app/views/entries/show.html.erb:
```diff
@@ -9,3 +9,5 @@
 <% end %>
 </ul>
+
+<a href=<%= edit_entry_path(@entry)%>>Edit Entry</a>
```
Double quotes around the href here?

christoomey commented on app/models/entry_form.rb:
```diff
@@ -1,66 +1,90 @@
+    entry_form = new
+    entry_form.entry = entry
+    entry_form.date = entry.date
+    entry_form.goals = entry.goals
```
Rather than setting these as state, can they be delegated?

christoomey commented on app/models/entry_form.rb:
```diff
@@ -1,66 +1,90 @@
-    goals_validations = goals.all?(&:valid?)
-    entry_validation && goals_validations
+  def entry
+    @entry ||= Entry.new(date: date, user: user)
```
I think we'll want to pass in the entry rather than initialize in here.

christoomey reviewed (commented):
I think we've pulled a little too much behavior into the form object here. Take a look at my comments, and likely worth chatting a bit in person, but I'm hoping we can accomplish the same goals with a good bit less code here.

johnschoeman committed:
Get third and final pass at edit entry feature working

This is another refactoring of the entry form object, which improves on
the previous refactoring and doesn't add new functionality.

This refactoring pivoted the role of initialize to take in the higher level
abstraction of an entry and only and entry, not the date or description
or goal_descriptions.  The setting of the actual form entries was pushed
to update_attributes for both entry creation and entry edit.

johnschoeman committed:
Refactor to use form partials for edit and new forms

No changes made to behavior with this, but uses the same form partial
with both views.
